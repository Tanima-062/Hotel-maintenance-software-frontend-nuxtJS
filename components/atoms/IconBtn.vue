<template>
  <button
    type="button"
    class="flex h-fit-content items-center justify-center"
    :class="styles"
    :disabled="disabled"
    @click="click"
  >
    <svg
      class="w-7 h-7 fill-current"
      :class="`${color} hover:${hoverColor}`"
      :style="`width: ${iconWidth}px; height: ${iconHeight}px`"
      viewBox="0 0 24 24"
    >
      <path :d="icon" />
    </svg>
  </button>
</template>
<script lang="ts">
import { Vue, Component, Prop, Emit } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  icon!: string

  @Prop({ default: 'text-white' })
  color!: string

  @Prop({ default: 'text-white' })
  hoverColor!: string

  @Prop({ default: false })
  disabled!: boolean

  @Prop({ default: false })
  round!: boolean

  @Prop({ default: '' })
  bgColor!: string

  @Prop({ default: '' })
  padding!: string

  @Prop({ default: 24 })
  iconWidth!: number

  @Prop({ default: 24 })
  iconHeight!: number

  @Prop({ default: '' })
  extraClass!: string

  private isAnim: boolean = false
  private interval: number | undefined

  get styles(): string {
    const classList = []

    classList.push(this.isAnim ? 'btn--click' : 'btn')
    if (this.disabled) {
      classList.push('cursor-not-allowed')
    }
    if (this.round) {
      classList.push('rounded-full')
    } else if (!this.extraClass.includes('rounded-sm')) {
      classList.push('rounded-md')
    }
    classList.push(this.bgColor ? `${this.bgColor} opacity-75` : 'opacity-85')
    if (this.padding) {
      classList.push(`${this.padding}`)
    }
    if (this.extraClass) {
      classList.push(this.extraClass)
    }
    return classList.join(' ')
  }

  private doAnim(): void {
    if (this.isAnim || this.interval) {
      return
    }
    this.isAnim = true
    this.interval = window.setInterval(() => {
      this.isAnim = false
      clearInterval(this.interval)
      this.interval = undefined
    }, 130)
  }

  @Emit()
  private click(e: Event): Event {
    this.doAnim()
    return e
  }
}
</script>
<style lang="sass" scoped>
.btn
  transition: opacity .1s ease-in
  &:hover
    transition: opacity .2s cubic-bezier(.4,0,.6,1)
    opacity: 1
  &--click
    transition: opacity .05s ease-out
    opacity: .7
.h-fit-content
  height: fit-content
</style>
