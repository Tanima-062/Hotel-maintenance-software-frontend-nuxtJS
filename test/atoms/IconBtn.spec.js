import { shallowMount } from '@vue/test-utils'
import IconBtn from '@/components/atoms/IconBtn.vue'

describe('Atoms IconBtn Check', () => {
  test('IconBtn Test', () => {
    const wrapper = shallowMount(IconBtn)

    // check if it is a vue element
    expect(wrapper.vm).toBeTruthy()

    // check Snapshot
    expect(wrapper.element).toMatchSnapshot()
  })
})
