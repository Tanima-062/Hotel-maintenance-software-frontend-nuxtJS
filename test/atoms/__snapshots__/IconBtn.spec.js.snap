// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms IconBtn Check IconBtn Test 1`] = `
<button
  class="flex h-fit-content items-center justify-center btn rounded-md opacity-85"
  type="button"
>
  <svg
    class="w-7 h-7 fill-current text-white hover:text-white"
    style="width: 24px; height: 24px;"
    viewBox="0 0 24 24"
  >
    <path
      d=""
    />
  </svg>
</button>
`;
